Create 0000-options-local-pragma.rst (#1)
pluralia/ghc-proposals: 0051-options-local-pragma.rst

Options Local Pragma
====================

.. proposal-number:: 
.. ticket-url::
.. implemented::
.. highlight:: haskell
.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/0>`_.
.. sectnum::
.. contents::

> We propose to add an opportunity to switch GHC-options locally. For instance, this code

Reviewer: “opportuniy” = chance, occasion

```suggestion
We propose to add functionality for switching GHC-options locally. For instance, this code
```

::

>  f = {-# OPTIONS_LOCAL -Wname-shadowing #-}

Reviewer: Might be a good idea to add:

```haskell
{-# OPTIONS_GHC -Wno-name-shadowing #-}
```

at the top, so that we do not rely on what warnings are switched on by default.

   let a = () in
   let a = 5 in
     a + a

> get warning ``-Wname-shadowing`` to line

Reviewer: ```suggestion
will result in a  ``-Wname-shadowing`` warning at line
```

::
 let a = 5 in

Motivation
------------

> There are the following reasons to implement such pragma:

Reviewer: ```suggestion
There are the following reasons to implement this pragma:
```

> 1) **Local control of warnings**

Reviewer: I have no clue how to use rst, but this formatting does not look right.

Author: I don't understand the problem. It looks like a bold text.

Reviewer: This appears to be a list with a single element consisting of this one bold line. But what you probably want to do is make the entire following text a list item.

Author: I didn't find how to create a list where one point consists of some paragraphs, but fixed it another way.

> To allow control warnings when you confident in your decisions without passing really important warnings.

Reviewer: “pass” is always intentional

```suggestion
To allow control of warnings when you are confident in your decisions without missing really important warnings.
```

> Warnings are compiler recomendations and often they can be really useful. It allows to avoid stupid mistakes and points to *weak* places in the code. You should follow it. But sometimes you can be confident in your locally code-decision and get some useless warning. It makes tracking warning list more difficult and you can pass really important warning.

Reviewer: ```suggestion
Warnings are compiler recomendations and often they can be really useful. They allow us to avoid stupid mistakes and point to *weak* places in the code. You should pay attention to them. But sometimes you can be confident in your local code-decision and get some useless warning. It clutters the warnings list and you can miss a really important warning.
```

 Consider two situations:
 

>  a) You want to enable warning locally (and it will be disables in all file).

Reviewer: ```suggestion
 a) You want to enable a warning locally (while it will remain disabled in the rest of the file).
```

 ::
  f = {-# OPTIONS_LOCAL -Wname-shadowing #-}
    let a = () in
    let a = 5 in
      a + a

>  b) You want to disable warning locally (and it will be enabled locally):

Reviewer: ```suggestion
 b) You want to disable a warning locally (while it will remain enabled in the rest of the file):
```

 ::

>   {-# OPTIONS -Wunused-do-bind #-}

Reviewer: ```suggestion
  {-# OPTIONS_GHC -Wunused-do-bind #-}
```

Author: I checked `OPTIONS`. Why is using `OPTIONS_GHC` more preferable?

Reviewer: > Previous versions of GHC accepted `OPTIONS` rather than `OPTIONS_GHC`, but that is now deprecated.

https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#options-ghc-pragma

 
  main :: IO ()
  main = {-# OPTIONS_LOCAL -Wno-unused-do-bind #-} do
    getLine
    return ()

> 2. **Using language extension locally**

Reviewer: ```suggestion
2. **Using language extensions locally**
```

> Using language extensions locally to avoid using it in extra places by a misatake.

Reviewer: ```suggestion
Using language extensions locally to avoid using it in other places by mistake.
```

> Language extensions allow to add useful things but "with great power there must also come great responsibility". As a result, sometimes it leads to a lot of mistakes and complicates code.

Reviewer: ```suggestion
Language extensions allow one to add useful things but "with great power comes great responsibility". As a result, sometimes it leads to a lot of mistakes and complicates the code.
```

Reviewer: “add useful things” is a little vague :)

Reviewer 2: I checked this phrase :) 
https://en.wikipedia.org/wiki/With_great_power_comes_great_responsibility

Reviewer 2: > “add useful things” is a little vague :)
I have no idea how to develop this topic.

Reviewer 2: I checked this phrase :)
https://en.wikipedia.org/wiki/With_great_power_comes_great_responsibility

Author: > “add useful things” is a little vague :)

I have no idea how to develop this topic...

Author: https://github.com/Pluralia/ghc-proposals/pull/1#discussion_r283257109

I checked this phrase :)
https://en.wikipedia.org/wiki/With_great_power_comes_great_responsibility

>   You can enable language extension locally (and it will be disabled in all file).

Reviewer: ```suggestion
  You can enable a language extension locally (and it will remain disabled in the rest of the file).
```

  ::
   x' = {-# OPTIONS_LOCAL -XPartialTypeSignatures #-}
     let a :: _
         a = ()
      in ()

Proposed Change Specification
-----------------------------

> Work with the GHC options is comfortable throw the special pragma (there is exist pragma ``OPTIONS``). We propose create a similar pragma OPTIONS_LOCAL which will do the same things but locally. You can see how it looks like in the section *Motivation*.

Reviewer: ```suggestion
GHC already support the ``OPTIONS_GHC`` pragma for configuring options for the file as a whole. We propose to create a similar pragma ``OPTIONS_LOCAL`` which will do the same things but locally. You can see what it looks like in the *Motivation* section.
```

Places for OPTIONS_LOCAL pragma:
 - expression
 - bind
 - declaration

Costs and Drawbacks
-------------------

1) **Estimate on development and maintenance costs**

Every compiler flag can require individual way to collaborate with ``OPTIONS_LOCAL`` pragma.

2) **Influence to learnability of the language**

> Similar like work with other pragmas.

Reviewer: But all pragmas are different. Maybe say here something about that pragma being _optional_ (ba dum tss) and non-essential for basic uses of the language. And after that it does not require any more learning after the `OPTIONS_GHC` pragma (well, this is not entirely true, as you still need to learn where and how to place it inside the file, somewhat like the `SCC` pragma).

Author: Rewrite this part.

3) **Remaining drawbacks**

None.

Alternatives
------------

> Pragma `OPTIONS` (but work globally).

Reviewer: `OPTIONS_GHC` itself it not  an alternative to this proposal. The alternative is “Do nothing”.

But are there really no other alternatives? I vaguely remember that we were discussing something at the very beginning of the project, those options might be worth mentioning in this section.

Author: I reread the beginning of the conversation and didn't find any mentions about alternatives.

Unresolved Questions
--------------------

> None.

Reviewer: Really? 🤔

Author: I don't know what I can write here. There are problems with implementation, but there are not problems with an idea or a design.

Author: Add something.
